ShareIt v.3

DpaNniK opened the pull request:
add item request
add test

DpaNniK committed:
ShareIt v.3
add item request
add test

DpaNniK committed:
ShareIt v.3
add item request
add test

DpaNniK committed:
ShareIt v.3
add item request
add test

DpaNniK committed:
ShareIt v.3
add item request

DpaNniK committed:
ShareIt v.3
add item request

DpaNniK committed:
ShareIt v.3
add item request

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/model/Request.java:
```diff
@@ -0,0 +1,28 @@
+@RequiredArgsConstructor
+@Entity
+@Table(name = "requests", schema = "public")
+public class Request {
```
Лучше назвать ItemRequest, а то в будущем могут появиться запросы чего еще либо и не будет понятно

DpaNniK replied:
done

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/model/Request.java:
```diff
@@ -0,0 +1,28 @@
+    private Integer id;
+    @NotNull
+    private String description;
+    private Integer requestorId;
```
Надо реализовать через связь ManyToOne

DpaNniK replied:
Сделал связь one-to-one

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/model/Request.java:
```diff
@@ -0,0 +1,28 @@
+    private Integer requestorId;
+
+    @Column(name = "created", nullable = false)
+    private LocalDateTime created = LocalDateTime.now();
```
Можно использовать @CreatedTimeStamp

DpaNniK replied:
done

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/mapper/RequestMapper.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Collection;
+
+@UtilityClass
+public class RequestMapper {
```
Можно использовать MapStruct

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/repository/ItemRequestRepository.java:
```diff
@@ -0,0 +1,19 @@
+    Collection<Request> getDistinctByRequestorIdOrderByCreatedDesc(Integer requestorId);
+
+    @Query("select r from Request r where r.requestorId <> ?1 order by r.created desc")
+    Page<Request> getDistinctForPagination(Integer requestorId, Pageable pageable);
```
Попробуй реализовать запрос средствами springData без Query

DpaNniK replied:
Написал такой запрос 
Page<ItemRequest> getDistinctByRequestorIdNotContainingOrderByCreatedDesc(User requestor, Pageable pageable);
Но почему-то метод стал дропаться с ошибкой 
No need to create transaction for [org.springframework.data.jpa.repository.support.SimpleJpaRepository.getDistinctByRequestorIdNotContainingOrderByCreatedDesc]: This method is not transactional.
Ковырял гугл, так и не понял, почему это происходит...

DpaNniK committed:
ShareIt v.3.1
Время created для ItemRequest определяется @CreationTimestamp
Сущность ItemRequest связана связью one-to-one с User

DpaNniK committed:
ShareIt v.3.1
Время created для ItemRequest определяется @CreationTimestamp
Сущность ItemRequest связана связью one-to-one с User
